### ISSUE TITLE
InsertPosition parent->child problem

### ISSUE THREAD

**Issue by kajzersoze** [xmonad/xmonad-contrib#551]
### Problem Description

I needed new window to appear on stack and not on master, so I've used InsertPosition, that works well but it has a problem, when parent window opens new window, that new window goes behind the parent window

It needs to go in front of the parent.

It does this 

![20210527-221805](https://user-images.githubusercontent.com/11412353/119891783-e9f9ee80-bf28-11eb-8b98-ba7fb078cde3.png)

instead of this 

![20210527-221852](https://user-images.githubusercontent.com/11412353/119891791-ed8d7580-bf28-11eb-832e-e3da38064596.png)

### Steps to Reproduce

`insertPosition End Newer` in managehook

`End` could be replaced with `Above` or `Below`, same thing happens, focus (`Newer` `Older`) doesn't affect this.

This example is with Galculator (opening Edit->Preferences), but any app would have the same problem.

**Comment by TheMC47** [comment]
Thanks for reporting! This is indeed an issue, and I think https://github.com/xmonad/xmonad/issues/300 addresses a fix for it: basically only apply `insertPosition` to windows that won't float.

You could use the code from there, wait for it to be included in a PR, or better yet, submit a PR yourself!

**Comment by liskin** [comment]
> basically only apply `insertPosition` to windows that won't float.

Isn't it a bit weird though that "End could be replaced with Above or Below, same thing happens, focus (Newer Older) doesn't affect this." ?

**Comment by liskin** [comment]
There's also [`isDialog`](https://hackage.haskell.org/package/xmonad-contrib-0.16/docs/XMonad-Hooks-ManageHelpers.html#v:isDialog) btw which could maybe help in some cases until https://github.com/xmonad/xmonad/issues/300 is dealt with

**Comment by kajzersoze** [comment]
> Thanks for reporting! This is indeed an issue, and I think [xmonad/xmonad#300](https://github.com/xmonad/xmonad/issues/300) addresses a fix for it: basically only apply `insertPosition` to windows that won't float.
> 
> You could use the code from there, wait for it to be included in a PR, or better yet, submit a PR yourself!

That works with windows that don't float , but it should work with any window, regardless of that.
I have Galculator btw as float :)

in dwm I've used [this](https://dwm.suckless.org/patches/attachabove/) for that, worked fine, simple patch it seems

**Comment by liskin** [comment]
> but it should work with any window

Well if you use `insertPosition` to open new windows after the currently focused one, and both of them are floating, then this is quite literally you asking for the window to be opened below. In xmonad, "after" and "below" are indistinguishable.

**Comment by kajzersoze** [comment]
It doesn't happen with `Above` , but the new window opens in master in that case.
The goal is to open a new window on stack.

**Comment by kajzersoze** [comment]
> There's also [`isDialog`](https://hackage.haskell.org/package/xmonad-contrib-0.16/docs/XMonad-Hooks-ManageHelpers.html#v:isDialog) btw which could maybe help in some cases until [xmonad/xmonad#300](https://github.com/xmonad/xmonad/issues/300) is dealt with

I just tried this, and it actually works with both float and tiled windows :+1: 

`fmap not isDialog --> insertPosition End Newer`

**Comment by kajzersoze** [comment]
Though there might be cases where this wouldn't work, I had recently an issue with some games for example, Rockstar made a new launcher for their games and I was waiting and waiting on some initial loader screen to finish, and it wouldn't finish so I killed it a few times and gave up, until I figured by pure chance that the loader was waiting because there was a dialog with a question, which was behind the initial screen, hence this issue.
Now, I can't test this one, at least not now, because for that I would have to delete all the config files for it and start over.
But this is better than nothing.

**Comment by kajzersoze** [comment]
I guess I've made a mistake in closing this issue.
While `isDialog` is not a bad solution, it still doesn't solve the situation with floating windows.
Now, maybe if there's a way to use `isDialog` for tiled windows (with insertPosition Below) and a `willFloat` (from another issue) for floating windows, in which case `insertPosition Above` would be used, that would be great, but I don't know how to do that combined.

**Comment by kajzersoze** [comment]
So I've decided not to use insertPosition and get used to the default, surprisingly enough this issue happens even without using insertPosition, I tried this many times and it's reproducible, in VirtualBox when you try to select an ISO image for Storage, and go for 'Choose a disk file' a file dialog is appearing behind the previous window, completely unvisible, in the picture I moved a window to show the one behind it, which should be in front

![20210608-170711](https://user-images.githubusercontent.com/11412353/121211610-26541580-c86c-11eb-82ea-79ee7023848a.png)

**Comment by kajzersoze** [comment]
The problem is that it depends on where you click it, focus is playing a big part, for example in this screenshot, if I click 'Choose a disk file...'  with my cursor on that yellow icon or where 'C' is, then the file dialog will open in front of it, but if the cursor is anywhere else right from it, starting from 'h', then it will go behind and become invisible.
 
![20210611-145618](https://user-images.githubusercontent.com/11412353/121689898-b8e6f580-cab4-11eb-85b7-00769d3c07f7.png)

**Comment by kajzersoze** [comment]
Well, with `focusFollowsMouse  = False` everything seems to be working fine (solves the problem in previous post), it makes things harder though, if one is not used to that routine, but it is what it is.
Placing new window shouldn't depend on focus and it should always be on top of the old window.

I'll close this, but this needs some attention, I know at least two people who gave up on Xmonad because of this particular issue (they didn't even know how to change the default master stack behaviour), I wouldn't either, had to ask on IRC to find that out, google didn't help, it took me some time to fight with some issues which followed after using `insertPosition`, I don't give up that easily :D

**Comment by liskin** [comment]
> but this needs some attention

I got a bit lost in all the discussion; do we have a minimal config and a bunch of simple steps that reproduce the issue? That would help us tremendously to give it the attention it needs.

**Comment by kajzersoze** [comment]
Understandable, it is confusing.
[Here's](https://pastebin.com/yBKgfVNU) my actual current config, it's not minimal but it's not that big either
For the problem with focus (last post with picture) , you don't even have to do anything, that happens with default setup, so no `insertPosition` is used, step is simple, on already existing or just created VM, go to Settings for that VM then Storage->Empty, you can see it in the picture, so if you click on 'Choose a disk file' with the cursor on the right part (starting from h) a file dialog will appear below that window, you wouldn't see it unless you move the old window.

Now the second problem, say you want to change the default `insertPosition Above Newer`, so you want for new window to appear on the right as opposed to the default left, basically new window to go to stack instead of master, you would then use `insertPosition Below Newer` , the problem is most obvious with floating windows, but it happens with tiled as well, for example in Firefox, if you save some file for example and then try to save it again, an overwrite dialog will appear but it wouldn't be visible, for floating windows I tried few and it's the same with any, let's say you set Galculator as floating, you open it, then go to Edit->Preferences and the dialog will appear behind (picture in first post)

That's it I guess, maybe I missed something, if I did I would be happy to help further.
I'll re-open this again, since there's a will to fix this.

**Comment by liskin** [comment]
Would you happen to have a reproducer that doesn't involve VirtualBox by any chance?

**Comment by kajzersoze** [comment]
For that problem no, found that by chance, it just so happens that the dialog with options you have to choose is outside of the focused window, hence that problem, anyway it's rare I guess and I don't know any other app like that, but there could be cases like that.

Also, I missed that in previous post, that issue requires this setting not to be set `focusFollowsMouse = False`, I'm not sure but I believe it defaults to `True` if not set

**Comment by kajzersoze** [comment]
I don't know why this relies heavily on focus, that shouldn't matter at all.
It's very simple really, just lose the focus part and place new window either on master or on bottom of the stack, make that configurable, always make the focus on new window, and if new window isn't tiled put it on top regardless of that configurable setting I mentioned before, so always use `insertPosition Above` in that case.

**Comment by kajzersoze** [comment]
Seems like there are several issues with the same problem open since 2016 lol
So never mind this one, who cares.
Well, it was a interesting month with Xmonad, it's really bizarre WM I must say, peculiar at least, good luck!

**Comment by liskin** [comment]
If you found other issues related to this one, would you please link them here? We still intend to investigate this, just haven't got to it yet (and yes, it might take more than a week). Any additional info helps.

**Comment by kajzersoze** [comment]
https://github.com/xmonad/xmonad/issues/106
https://github.com/xmonad/xmonad/issues/138

I found those 5 days ago, would have to look in history for more, can't even remember what exactly I was looking for, something in config or in haskell, doesn't matter
btw it's not that I'm in a hurry or something, I just moved away
